Objects/typeobject.c

```diff
@@ -5961,14 +5961,26 @@ hackcheck(PyObject *self, setattrofunc func, const char *what)
         return 1;
     }
     assert(PyTuple_Check(mro));
-    Py_ssize_t i, n;
-    n = PyTuple_GET_SIZE(mro);
-    for (i = 0; i < n; i++) {
+
+    /* Find the (base) type that defined the type's slot function. */
+    PyTypeObject *defining_type = type;
+    Py_ssize_t i;
+    for (i = PyTuple_GET_SIZE(mro) - 1; i >= 0; i--) {
         PyTypeObject *base = (PyTypeObject*) PyTuple_GET_ITEM(mro, i);
-        if (base->tp_setattro == func) {
-            /* 'func' is the earliest non-Python implementation in the MRO. */
+        if (base->tp_setattro == slot_tp_setattro) {
+            /* Ignore Python classes:
+               they never define their own C-level setattro. */
+        }
+        else if (base->tp_setattro == type->tp_setattro) {
+            defining_type = base;
             break;
-        } else if (base->tp_setattro != slot_tp_setattro) {
+        }
+    }
+
+    /* Reject calls that jump over intermediate C-level overrides. */
+    PyTypeObject *base = defining_type;
+    while (base && base->tp_setattro != func) {
```

Comment:
I wonder if this would read better if you wrote it as
```
for (PyTypeObject *base = defining_type; base != NULL; base = base->tp_base) {
    if (base->tp_setattro == func) {
        /* appropriate comment (same as old code?) */
        break;
    }
    ...
```
(That for-loop is IMO a more idiomatic way to iterate over a linked list.)